freshmaker/lightblue.py

```diff
@@ -1579,5 +1579,239 @@ def _get_images_to_rebuild(image):
         directly_affected_nvrs = {
             image.nvr for image in images if image.get("directly_affected")
         }
+        # Some images that aren't marked as directly affected may have already been fixed
+        # in the latest published version of the image. Use those images instead.
+        self._filter_out_already_fixed_published_images(
+            to_rebuild, directly_affected_nvrs, rpm_nvrs, content_sets
+        )
+
         # Now generate batches from deduplicated list and return it.
         return self._images_to_rebuild_to_batches(to_rebuild, directly_affected_nvrs)
+
+    def _filter_out_already_fixed_published_images(
+        self, to_rebuild, directly_affected_nvrs, rpm_nvrs, content_sets
+    ):
+        """
+        Replace images in ``to_rebuild`` that are not directly affected and have published fixes.
+
+        When an image and its parents in ``to_rebuild`` are not directly affected, it's possible
+        that the image had been rebuilt outside of Freshmaker and published with the fix applied.
+        In this case, Freshmaker should not rebuild the image and its parents. The latest published
+        image is found using the pull spec of the parent image from the perspective of
+        the last directly affected image (FROM line).
+
+        :param Iterable to_rebuild: the list of images to rebuild
+        :param Iterable directly_affected_nvrs: the set of image NVRs in ``to_rebuild`` that are
+            marked as directly affected
+        :param Iterable rpm_nvrs: the list of RPM NVRs with the fixes in the advisory
+        :param Iterable content_sets: the list of content sets that the RPMs in ``rpm_nvrs`` are
+            released in
+        """
+        for image_group in to_rebuild:
+            # Find the first index in image_group of an image that is not directly
+            # affected with parents that are also not directly affected
+            not_directly_affected_index = None
+            # Skip the first image in the group since it is always directly affected
+            for i, image in enumerate(image_group[1:]):
+                if image.nvr in directly_affected_nvrs:
+                    not_directly_affected_index = None
+                elif not_directly_affected_index is None:
+                    # We must add 1 since the first index was skipped in the for loop
+                    not_directly_affected_index = i + 1
```

Comment:
I think it would be possible since the base image could have the RPM installed, the intermediate image uninstalls it, and the child image reinstalls the same RPM. It's very unlikley but it's not very wasteful to guard against this.

What do you think?